### DISCUSSION TITLE
Nextjs Link unreliable on production only

### DISCUSSION THREAD

**Discussion by VercammenJens** [vercel/next.js#36670] [Help]
Hi,

I'm not sure if this belongs inside Issues or Discussions, but I'll post this here first.

We are currently facing an odd issue in our production website which is hard to reproduce and seems very unreliable. Our application, next 12.1.0 with a Strapi headless CMS, has a blog-like section with landings page, category pages and post pages. In the part of the website we are using the content which is fetched from Strapi using SSG. We are using the [On-demand Revalidation](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation-beta) to update the content of the post pages and a 10 seconds revalidation on the other pages due to not being able to revalidate multiple pages on demand that contain the same or a part of the same content.

From time to time going from a parent page to a dynamic sub page inside this blog section redirects the user to a 404 page, due to the next/link's `Link` tag giving an error on the prefetch (404 not found). For example, going from landings page to green energy category. Once you hover over the Link the prefetch starts, and fails. When clicking on this Link you get redirected to a 404 page, but upon refresh the page is correctly loaded.

Inside these dynamic sub pages I am utilizing the `return { notFound: true }` inside the `getStaticProps`.

Is there anyone who has had a similar issue or knows a fix to this weird "dead" link, which is not actually "dead"? Going to test it without the `return { notFound: true }` while waiting for replies.

Thanks!

here's some code examples:

**Landings page SSG**
```typescript
export const getStaticProps = async ({ locale }) => {
  try {
    // fetch Articles & Categories
    const [content, categories, articles] = await Promise.all([
      getEncyclopediaLandingsPage(),
      getEncyclopediaCategories(),
      getEncyclopediaArticles()
    ])

    return {
      props: {
        articles,
        categories,
        content,
        ...(await serverSideTranslations(locale))
      },
      revalidate: 10
    }
  } catch (error) {
    return {
      props: {
        articles: [],
        categories: []
      },
      revalidate: 10
    }
  }
}
```

**Category page SSG**
```typescript
export const getStaticProps = async ({ params, locale }) => {
  const { category: categorySlug } = params

  const articlesQuery = stringify({
    locale: Language.DUTCH,
    filters: {
      subCategory: {
        category: {
          slug: {
            $eq: categorySlug
          }
        }
      }
    }
  })

  const subCategoriesQuery = stringify({
    locale: Language.DUTCH,
    filters: {
      category: {
        slug: {
          $eq: categorySlug
        }
      }
    }
  })

  // Axios get calls to get content from Strapi headless CMS
  const [category, subCategories, relatedArticles, allArticles] = await Promise.all([
    getEncyclopediaCategoryBySlug(categorySlug),
    getEncyclopediaSubCategoriesByQuery(subCategoriesQuery),
    getEncyclopediaArticlesByQuery(articlesQuery),
    getEncyclopediaArticles()
  ])

  if (!category) {
    return {
      notFound: true
    }
  }

  return {
    props: {
      allArticles,
      categoryContent: category,
      relatedArticles,
      subCategories,
      ...(await serverSideTranslations(locale))
    },
    revalidate: 10 // re-generate page when a request comes in, once every 10 seconds
  }
}

export const getStaticPaths = async () => {
  const paths = []

  const categories = await getEncyclopediaCategories()

  categories.forEach((cat) => {
    const categoryPaths = [
      { params: { category: cat.slug }, locale: 'nl' }
      // { params: { category: cat.slug }, locale: 'fr' }
    ]
    paths.push(...categoryPaths)
  })

  return {
    paths,
    fallback: 'blocking' // will server-render pages on-demand if the path does not exist
  }
}
```

**Link from landings page to category page**
```typescript
<Link key={category.id} href={routes(language, category.slug).encyclopediaCategory} prefetch={false}>
  <a className={...}>
    ....
  </a>
</Link>
```

**Comment by icyJoseph** [comment]
One thing to be mindful about, is that in `notFound` you should also return revalidate! otherwise the error page would be cached for a long time!

```js
 if (!category) {
    return {
      notFound: true, 
      revalidate: 1000, // or some other long appropriate time
    }
  }
```

I think, if you want to prevent failures in Promise.all or any other async work on category, from generating 404, you should just let the error bubble up (aka re-throw it), according to this docs, https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#error-handling-and-revalidation.

I imagine that, if the request that gets a category fails, you gracefully handle it to return a falsy value, and that tell Next.js to actually replace the cached page with a 404. 

Still, why would it work fine when you refresh the page. Prefetching fails, but works when reloading, **then we should think that the cached version of the page still exists**, but sending the json payload to render on the client failed. Do you have any server side logs?

Perhaps you could wrap the entire category GSP with a try/catch, log the issue, and throw an error (or just re-throw).

**Reply by icyJoseph** [reply]
> This JSON file will be used in client-side routing through [next/link](https://nextjs.org/docs/api-reference/next/link) or [next/router](https://nextjs.org/docs/api-reference/next/router). When you navigate to a page that’s pre-rendered using getStaticProps, Next.js fetches this JSON file (pre-computed at build time) and uses it as the props for the page component. This means that client-side page transitions will not call getStaticProps as only the exported JSON is used.

I am now thinking that this might be a race condition? Where the ISR has replaced the page (HTML + JSON), at the time the prefetch request has arrived, so there's nothing to serve (404)?

**Reply by VercammenJens** [reply]
Thanks for the tip about the notFound needing a revalidation aswell. I did not know this! The fact prefetching fails, but the reload does load the page makes no sense to me atm. Cache might explains this, but is there a reason to test/see if it's a cached page or not?

We do have a bugsnag for some errors, but currently don't have a server side logs tracker. We use vercel so the build logs are visible. Guess I'll attach a bugsnag error to the GSP try/catch and hopefully get more information about the issue.

**Reply by VercammenJens** [reply]
> Where the ISR has replaced the page (HTML + JSON), at the time the prefetch request has arrived, so there's nothing to serve (404)?

It being a race condition could be possible indeed, but isn't the purpose of the prefetch to run through the getStaticProps and "wait" until the async is done?

**Reply by VercammenJens** [reply]
The biggest issue is that it's still unclear how to reproduce this, and that it only happens on production. So pretty difficult to debug. Usually I try to apply a fix, and then it is possible that I have to wait a couple of days for it to happen again.

**Reply by icyJoseph** [reply]
> It being a race condition could be possible indeed, but isn't the purpose of the prefetch to run through the getStaticProps and "wait" until the async is done?

Well, I feel like I am throwing a wild guess here, but for sure the prefetching just gets a JSON file, already associated with the page. The race condition would be caused by ISR replacing that JSON (and HTML) file with a new file, and the prefetch request arriving exactly at the same time. Extremely unlikely, but after reading the docs and your description it is the only thing that comes to mind.

Feels like more people should report it, given the amount of people that use Next.js

**Comment by kendallstrautman** [comment]
To test locally, you can try running `next build && next start` and you may see the production behavior you're looking for. The prefetching behavior only occurs on production. 

I'm seeing similar behavior in that a clientside redirect throws an unexpected error but on refresh the expected view renders. In my case I _do_ expect the link to navigate to a 404 view (context: linking to versioned docs that may not exist many versions back), but instead I get a clientside error, which catches in the app level error boundary. Even though in the server I see how the 404 status is calculated properly. 

If I hard-navigate to the route, the proper 404 view renders, so this does feel like race condition behavior to me, and very difficult to debug.

**Comment by imjared** [comment]
We're seeing what sounds like similar behavior. It's very intermittent and no one on the team has been able to reproduce it for weeks. We're on the pages setup with `getServersideProps` attached to all pages. It acts mainly as the authentication gateway but also occasionally fetches data.

We've seen in our RUM setup that when users experience this issue, the URL successfully changes but the page transition never actually starts. No network request is initiated in the browser. Upon hitting refresh, the page will load immediately. Additionally, it seems that when users get into this state, it's hard to get out of until they have refreshed their browser.
